Check dynamo graph-breaks in CI

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96480
* __->__ #96346

- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

Refactor graph-break CI check

Take steps toward merging checker with existing check flow,
consider merging it all the way inside the bench runner.

csvs

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Check dynamo graph-breaks in CI

- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

[ghstack-poisoned]

ezyang commented on benchmarks/dynamo/ci_expected_accuracy/inference_huggingface.csv:
```diff
@@ -0,0 +1,43 @@
+dev,name,batch_size,accuracy,calls_captured,unique_graphs,graph_breaks,unique_graph_breaks
```
It's not obvious to me that we actually want all these columns here? calls_captured in particular feels likely to wobble, in ways that we ought not to care about?

wconstab replied:
i thought about this, and it seems easier to check in the exact csv that is easy to produce using our current flows.

and to your point about wobble, we just don't check the other fields- we check "graph_breaks"

wconstab replied:
i think the cleanest would be 
1) the msg also tells you "run this command: pytorch/benchmarks/dynamo/update_expected.py --pr {PRNUM}"
2) there is some clever mapping so the script can download the right .zips, and extract the csvs (2 csv per zip currently)
3) it can also spit out newly stripped CSVs that only contain the cared-about columns
4) and 'git add' the changes so you don't have to 

But i have to figure out (2), not sure if it is trivial or we have to do something to make it possible.  CC @ZainRizvi any help here?  (best practices for (a) how to interact with artifact files from a script (b) any stable 'ENV' inside test jobs that i can use to build a URL to download artifacts etc

desertfire replied:
For (2), you need an action after the test runs finish (see https://github.com/pytorch/pytorch/pull/95675 as an example). I am ok with doing this as a followup PR.

ezyang commented on benchmarks/dynamo/check_graph_breaks.py:
```diff
@@ -0,0 +1,75 @@
+        msg += textwrap.dedent(
+            f"""
+        If this change is expected, you can update `{args.expected}` to reflect the new baseline.
+        This can either be done manually, or by downloading artifacts from your PR CI job."
```
what's the name of the artifact in HUD would be useful here

ezyang commented on benchmarks/dynamo/ci_expected_accuracy/inference_huggingface.csv:
```diff
@@ -0,0 +1,43 @@
+dev,name,batch_size,accuracy,calls_captured,unique_graphs,graph_breaks,unique_graph_breaks
+cuda,AlbertForMaskedLM,1,pass,439,1,0,0
```
Something I did not realize until seeing this, is that we get a csv per shard. This means that I have to click through and download *eight* times to get all of them. Maybe we should have a script that downloads all the csvs from a job...

wconstab replied:
yea, i am wanting to build more tooling for this too. i have some other ideas...

ezyang commented:
Thanks for doing this!

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented on .ci/pytorch/test.sh:
```diff
@@ -332,10 +332,15 @@ test_dynamo_benchmark() {
-
+    python benchmarks/dynamo/check_breaks.py --actual \
+      "$TEST_REPORTS_DIR/inference_$suite$shard_id.csv" \
+      --expected "benchmarks/dynamo/ci_expected_accuracy/inference_$suite$shard_id.csv"
```
> Why not just put this in test_single_dynamo_benchmark

I can try that.  I was trying to only run the checker in the non-perf jobs but it looks like i can still do that from in there.

> a downside to having it be a separate script is if there is an unrelated failure, you don't get the graph break stats either

can you say more here? iiuc the job will still run all the models first (and that prints graph break stats model by model as it goes),  then the csv is saved (so you can always access it) at that point, and finally the check runs, which is the last thing .... *OH* do you mean, run the 2 checks (train/infr) after running the sweeps for train/infr?

i'm also confused about ` if there is an unrelated failure` -- if the bench runner itself bails out, i'm not making things worse am i?

ezyang replied:
So our batch scripts run with -e, which means after the first one fails the second one never gets run. So the antipattern here is, eg rexnet flaky fails accuracy, we stop running, and you don't realize there are ALSO graph break mismatches

wconstab replied:
alright, the thing i was missing is the bench runner keeps running and reports fails at the end, so i could integrate the graph break checks with the runner too.  makes sense, ill see if its hard

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
if you are out of patience and just want to land something, holler and I'll probably approve you to move things along

wconstab commented:
> if you are out of patience and just want to land something, holler and I'll probably approve you to move things along

i actually want to land it together with the WIP (next PR, which adds a script for downloading all the artifacts for the job) next week.  Anyway i'm on PTO tmrw so i won't rush this in.  It's probably landable though, if you're wanting it urgently, i can update the csvs monday and push it.

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

Refactor graph-break CI check

Take steps toward merging checker with existing check flow,
consider merging it all the way inside the bench runner.

csvs

[ghstack-poisoned]

wconstab commented:
ok @ezyang if you feel inclined (and this passes CI) you can stamp and mergebot, otherwise i can take care of it next week.  I have gotten the updater tool (next PR) in an almost landable state so i am not too worried about the inconvenience of updating the CSVs in the interim.

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

Refactor graph-break CI check

Take steps toward merging checker with existing check flow,
consider merging it all the way inside the bench runner.

csvs

[ghstack-poisoned]

ezyang commented:
By the way, for the differential between dynamic and non-dynamic, my preference is not to have a completely separate dynamic csv, and instead a hand-written list of "we have more graph breaks here".

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

Refactor graph-break CI check

Take steps toward merging checker with existing check flow,
consider merging it all the way inside the bench runner.

csvs

[ghstack-poisoned]

wconstab commented:
> By the way, for the differential between dynamic and non-dynamic, my preference is not to have a completely separate dynamic csv, and instead a hand-written list of "we have more graph breaks here".

what would the list contain, names of models and deltas compared to static?

In any case I planned to land this as is and probably let you tackle the dynamic part.  Or at least let me come back to it next week.

wconstab committed:
Update on "Check dynamo graph-breaks in CI"


- add graph-breaks baselines
- add check_graph_breaks script (message users on regress or improvement)
- hook up test.sh for existing accuracy job

Refactor graph-break CI check

Take steps toward merging checker with existing check flow,
consider merging it all the way inside the bench runner.

csvs

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

ezyang commented:
> what would the list contain, names of models and deltas compared to static?

For the most part, I'd expect most models to have the same number of graph breaks with/without dynamic, so it would be a list of one or two models that this is not true for.

wconstab commented:
@pytorchbot merge

wconstab commented:
> This was missed because periodic wasn't included in this PR. It looks like a forward fix is needed? Or feel free to revert if you need time to look into this

I will shoot for a forward fix.  Ping me if i'm taking too long.
